### ISSUE TITLE
DC-16 Altitude wont show in Log

### ISSUE THREAD

**Issue by helmarw** [DevFor8/Mav2Duplex#4]
Hi, i think found a strange bug?!

The altitude is shown on the jetibox and jetibox emulation on my DC-16
but the sensor for Altitude doesnt show anything, hence not getting saved in the logfile. 
see picture (ALT: 0.8m but beneath Mav2Dup ALt: -)
![dc_16_alt_bug](https://user-images.githubusercontent.com/10408121/28492259-0b775624-6f00-11e7-90b9-dd3f54adcae7.jpg)

i tried the most recent version too. Didnt change anything. 
I looked through the code but i cannot find the culprit.
On the other hand voltages and currents are being shown in the jetibox and as separate sensor (voltage is shown in the picture as an example)
and also saved in the logfile. So im pretty sure its some glitch.
Maybe someone can point me in the right direction

Also Lat and Lon (and a few others i removed from the logfile) do not get saved in the logfile
here you can see a summary of the sensors im saving during flight
![screenshot 2017-07-22 17 09 13](https://user-images.githubusercontent.com/10408121/28492347-c7dc0b24-6f01-11e7-9cb9-7fa69b1aab7b.png)

as you can see all sensors have 1475 items, except alt, lon and lat only 1 (sometimes its also a few more but never more than 10 or so.

Best
Helmar

**Comment by 0n3-70uch** [comment]
Hi Helmar, 

Is it possible that you are using an Arduino with 3,3V and 8MHz? I'm working on implementing the code into the ArduCopter source code. So no more device...

Greetings
Fabi

**Comment by mtbsteve** [comment]
Helmar,
You may look into my repository 
https://github.com/mtbsteve/APM2EX and compart. 
I forked the code for the Jeti telemetry a while ago. The EX telemetry including alt and lat/long works well with the latest Arducopter 3.5 version. 
Cheers Stephan

**Comment by helmarw** [comment]
@0n3-70uch: no i dont, im using an Arduino Nano 5V / 16MHz
otherwise it would not work at all, all other sensors with the exceptions you can see above are working

**Comment by helmarw** [comment]
i forked the source myself and working on that one, since i changed all of the screens in the jetibox according to my needs: https://github.com/helmarw/Mav2Duplex
i found a rather curious workaraund:
switched batt % with Alt, since Alt did not get sensor data, and i find Batt % rather useless
in fact everything higher 6 does not get evaluated as sensors data in EX sensors, except HDOP, GPSFix and GPSSats,  which doesnt make much sense, i cant figure out why, yet

so now the sensor Alt is in position 2 (former Batt %) and gets displayed and saved in logfile as it should be. So my wild guess there is something wrong in the library JETI_EX_SENSOR

**Comment by helmarw** [comment]
ok, here comes the test from this morning. as i said before i switch the sensor position Bat% and Alt.
now Alt gets the data but Bat% not (compare with above). I reseted all the sensor on my DC-16 before and logged all sensors now.
<img width="480" alt="bildschirmfoto 2017-07-24 um 07 40 52" src="https://user-images.githubusercontent.com/10408121/28509684-2187b694-7044-11e7-935b-bc048d36965f.png">

now you see that sensors 6-11 do not get any useful data, but "Alt" at position 2 does now
and then sensors 12-15 getting data again but not the same amount, which i do not understand yet

**Comment by helmarw** [comment]
i did a second test on the ground. switched sensor 12 (HDOP) with sensor 6 (Climb)
Climb is getting data now, but HDOP is not. so there is clearly an issue with the evaluation of the sensors
6 to 11
can one just comment out the sensors 6 to 11 ?! leaving the order as it is, so 1..5 and 12..15 
or does the gap form 6..11 in between cause other problems ?!

**Comment by helmarw** [comment]
@mtbsteve: doesnt work for me, cannot compile, and you removed the jeti box and from the readme also the Altitude. tried to use your libraries vice versa no luck...

**Comment by helmarw** [comment]
@0n3-70uch: how will you implement it in ArduCopter. Was wondering how the user will be able to configure their JetiBox screens. Not everyone has a DC-16 or later models.
Just the EX-Sensors will not cut it im afraid

**Comment by RadekVoltr** [comment]
I just released new version of code. It is masive change so I renamed Arduino project. Tested with DC-16 and all data is receiving correctly. Please test my version first and if you don't find issues, then you can add what do you want. Also there is support for Mega32u4 / Leonardo devices with USB port (better for debug and uploads as you don't need to reconnect and you can use USB directly) and lot of memory improvements. So I believe that there will be more space for additional old telemetry screens. I don't have problem to add additional screens pullrequest to code so don't hesitate to add or improve code. Standalone Jetibox library will go out later.

**Comment by helmarw** [comment]
k, thnx will do
only problem is, do i need to remove the resistor ?!
or can i test it just the same with it ?!

**Comment by helmarw** [comment]
sorry first issue, i cannot built it. 
took the latest arduino 1.8.3, in your readme you are station compatible to 1.8.6, but i could not find this version.
Here are the error msgs im getting
Arduino: 1.8.3 (Windows 10), Board: "Arduino Nano, ATmega328"

Archiving built core (caching) in: C:\Users\helmar\AppData\Local\Temp\arduino_cache_197849\core\core_arduino_avr_nano_cpu_atmega328_78895938c5cc955e38175117f4c7b56b.a
C:\Users\helmar\AppData\Local\Temp\ccvY1eta.ltrans0.ltrans.o: In function `setup':

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/Mav2EX.ino:235: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/Mav2EX.ino:235: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\AppData\Local\Temp\ccvY1eta.ltrans0.ltrans.o: In function `loop':

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/Mav2EX.ino:539: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/Mav2EX.ino:539: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/Mav2EX.ino:610: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\AppData\Local\Temp\ccvY1eta.ltrans0.ltrans.o:C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/Mav2EX.ino:610: more undefined references to `mavlink_comm_0_port' follow

C:\Users\helmar\AppData\Local\Temp\ccvY1eta.ltrans1.ltrans.o: In function `request_mavlink_rates':

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:52: undefined reference to `mavlink_system'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:52: undefined reference to `mavlink_system'

C:\Users\helmar\AppData\Local\Temp\ccvY1eta.ltrans1.ltrans.o: In function `read_mavlink(int)':

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:77: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:77: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:78: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:78: undefined reference to `mavlink_comm_0_port'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:91: undefined reference to `mavlink_system'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:92: undefined reference to `mavlink_system'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:110: undefined reference to `mavlink_system'

C:\Users\helmar\Documents\Arduino\Mav2Duplex-master\Mav2EX/MAVLink.ino:111: undefined reference to `mavlink_system'

collect2.exe: error: ld returned 1 exit status

exit status 1
Error compiling for board Arduino Nano.

**Comment by helmarw** [comment]
regarding the former version: theproblem is actually this commented code
/* the following parameters are not used by me
#define ITEMNAME_6 F("Yaw")
#define ITEMTYPE_6 F("Deg")
#define ITEMVAL_6 (short*)&osd_yaw
#define ITEMNAME_7 F("Pitch")
#define ITEMTYPE_7 F("Deg")
#define ITEMVAL_7 (short*)&osd_pitch
//#define ITEMVAL_7 (short*)&debug1
#define ITEMNAME_8 F("Roll")
#define ITEMTYPE_8 F("Deg")
#define ITEMVAL_8 (short*)&osd_roll
//#define ITEMVAL_8 (short*)&debug2
*/

which does not do anything to the lines starting with #define.
i realized that because im getting yaw, pitch and roll values on the screen in the jetibox emulation
that should not have worked because the lines where supposed to be commented
i simply removed the lines 6 to 8 (climb, Lat and Lon) and put back the original ones (Yaw, Deg and Roll)
now its working, at least im getting the same amount of data for all sensors. Have to do a field test still
Lat and Lon i anyway have on the pixhawk logfile so its not of utmost importance  on the DC-16
i know where im flying since i see the copter still ;)
Climb rate ill but instead of Batt% i think

**Comment by RadekVoltr** [comment]
Purpose of GPS on DC-16 is copter localization in case of fly-away. System should be able deliver 16 values so you can play with them as you want. I like to see roll/pitch before start to detect anomalies in IMU startups but it is up to you. New version have slightly more understandable way of value definition in code so changes should be also easier

**Comment by RadekVoltr** [comment]
There is new version of source code with right GSCMav library, I forget to copy it from Libraries in Arduino into Libraries in source tree. Sorry about that, should compile ok. Also Arduino version looks like typo.

**Comment by helmarw** [comment]
k thnx, its compiling now.
what about the resistor, can i leave it there for testing ?
or wouldnt work then ?

**Comment by RadekVoltr** [comment]
You can but you have to change jetibox init to work with separated tx,Rx. Check header file

Sent from my Windows 10 phone

**Comment by helmarw** [comment]
btw im getting warnings if i use german as language for the IDE
if i change it to english everything is fine, i can switch back and forth 
strange thing, now its gone
still have to test it though

**Comment by helmarw** [comment]
sorry two issues, update speed is even slower than before, e.g. to change mode its take about 6-7s
before it shows. On the "old" one did it in half as much, still slow thought... same with other 
parameters which can change quickly e.g. (Rol, Yaw and Pit)
msgs do not show except "Mav2EX init ok"
but usually when i dont have 3D-fix and i try to arm with Posh i get a prearm error
but here nothing...
im using a second arduino nano for for testing, so i can directly compare them

**Comment by 0n3-70uch** [comment]
Hello,

thanks for the nice update. Here my short summary:
- (+) more stable now
- (+) working with REX receiver (better timing)
- (+) all parameters visible
- (+) no resistor... a bit less soldering :-)
- (-) not working with 3,3V 8MHz Arduino Pro Mini
- (-) update rate is low

Is there any way to get it working on 8MHz CPUs? 

@helmarw Yes my plan is an implementation. But at the moment only with EX sensor data. I'm in a good contact with Martin (main developer of JETI)... I'm testing the new JETI Assist receivers at the moment and I wanna have the same configuartion for Arducopter. So the problem is not the .bin file but the time... I need more time... argh... ;-).

An interesting good working libary for JETI but only with Hardwareserial:
https://sourceforge.net/projects/jetiexsensorcpplib/

Greetings
Fabi

**Comment by RadekVoltr** [comment]
Hi

Speed issue is weird. Can you give me some details for faster fix ?

- Hw used ?
- Resistor or single wire ?
- Serial speed and all SR params from apm ?
- Serial is shared with osd or other uC ?
- Update is slow on all params ?
- Speed is same on old screen and ex data ?

Thanks
Radek 

Sent from my Windows 10 phone

**Comment by 0n3-70uch** [comment]
Hi Radek,

- HW: Arduino Pro Mini 5V 16MHz Atmega328P and Pixhawk.
- Single Wire
- 57600
- no not shared
- as I can see: yes
- yes. Changing between pages (jetiibox) is fast only the values are slow.

Hope this is helpful for you...

**Comment by helmarw** [comment]
- HW: Arduino Nano 5V 16MHz Atmega328P and Pixhawk.
- Single Wire
- 57600, hooked on Telem2, SR2 params all on 2 except SR2_PARAMS=0, but i can check again
- no not shared (for testing) but also shared with OSD (only TX) doesnt seem do make a difference
- as I can see: yes
- yes. Changing between pages (jetiibox) is fast only the values are slow.

thnx
Helmar

**Comment by RadekVoltr** [comment]
I am little busy right now and not sure when I will find time to check it more deeply. 

Can you try to configure your serial in same way as library request itself : 
SRx_EXT_STAT 4
SRx_EXTRA1 4
SRx_EXTRA2 4

and rest on zeros ? Also you can try to increase serial speed to 115K, it could help. As screens switch fast enough it looks like problem with Mavlink. Mavlink can flood uC with data and due relativelly small serial buffer in Arduino it could lead to problem with colisions and non complete packets.

**Comment by helmarw** [comment]
will give it a try, but a minimOSD works very well with the parameters im using, no lag at all. minimOSD uses the same Atmega328P, so i doubt that its the serial buffer is the only root cause, but i cant be sure.
In parallel im working on my code to speed things up, maybe i find something in the process.
apparently you can increase the fifo buffer from 64byte to 256bytes, but looks a bit ugly to do so in my opinion

**Comment by RadekVoltr** [comment]
minimOSD is completelly different issue - it uses FastSerial (or SingleSerial, depends on version) which is not compatible with new hardware like Leonardo (and lot of people asked for that). Also minimOSD don't need to do sending data to jeti which is blocking process. Right now I am going to test new version of code with callback function which will allow read mavlink much often and I hope that it will help.

**Comment by RadekVoltr** [comment]
Good news to everyone :-D It looks that I managed FastSerial library working on Mega32u4 hardware. It takes more memory as there is still original USB and new Fast but it is not big deal and we should fit during Debug. On Mega328 is target to have FastSerial only and I will try to manage removal of USB serial without Debug active. Now I am checking custom version of MAVLink used by MinimOSD Extra by Nightghost which maybe have shared buffer which allow us use one buffer for Mav and Serial. So then can be buffer bigger and so we will not loose data on the way. Current issue is racing condition between JETI and MAVLink because it looks that processing MavLink during JETI sending breaks protocol (probably there is some timeout between characters in frame). I will see if FastSerial/Mavlink with bigger buffer will help but fullsize Jeti packet takes lot of time (roughly about 80ms, depends on what is in) and you can loose lot of data from mavlink during that time. I will try also to keep interrupts working and other optimizations.

**Comment by helmarw** [comment]
regarding the SR# prameters:
SR#_ADSB apparently is necessary to get the LastMessage, if i put it on 0 no msgs at all
SR#_RC_CHAN i put on 1 since i need it to switch the OSD panels
if i put any of the values on 4, then the jetibox emulation resets very often
anyway i think updates twice every second is enough, but still it takes at least 2-3seconds
to update (with the "old" code), so i dont think increasing the update rate will help much

those are my parameters im using atm:
![screenshot 2017-07-25 19 16 26](https://user-images.githubusercontent.com/10408121/28584651-c755df9a-716d-11e7-8f7a-1bb2fb64bd36.png)

**Comment by helmarw** [comment]
@RadekVoltr: sounds great, looking forward to test it ;)

**Comment by 0n3-70uch** [comment]
@RadekVoltr Just one word: Nice :-)

**Comment by RadekVoltr** [comment]
yes, fastserial was used but removed in new version due Mega32U4 issues
ad params - code is clearly missing packets. So sending less packets but more often increase chance that packet will be captured

**Comment by helmarw** [comment]
why should i add params ? or you referring to something else ?

just sending a lot of packets and hope they get captured ?!
like shooting on a sparrow with gatling gun and hope you hit it ;)))
i prefer the sniper version ... one bullet, one headshot - metaphorically speaking of course ;)

**Comment by RadekVoltr** [comment]
New version released
- FastSerial and GCS changes (BEWARE : Code is not compatible with standard FastSerial)
- increased TX buffer
- MAVLink messages parsed during sending JETI data
- Interrupts allowed during soft serial processes
- few minor fixes for speed increase on JETI sending side

As result
- able parse mavlink messages 4x per second without loosing data
- able to send data slightly faster (10%) due Jeti sensor code optimization

This is maximum desirable speed with current set of sensors / data. Sending each jeti packet takes about 110ms and I have to send 5 of them to be sure that all data is delivered. So there is about 500ms lag due jeti protocol specs (9600/9/2 serial, 20ms wait for response, limited size per frame, text info required in all data) and it looks that there is also some lag on DC-16 side. It is clearly visible with Pitch/Roll data as Ex item and on Jetibox display. Jetibox display is faster because is included in each frame but Ex data only in some.

So from this side, I think that current speed is maximum which I can get. I am thinking about implementation of kind of "priority" items for Mode item so reaction could be faster (they will be included in all frames so rest of data will be slowed but it is not big deal as they are less important) or , if you want to have data faster, just comment items which are not usefull for you (especially long items like GPS can speed up updates). 

Please test it and let me know, I tested only on Mega32U4 device but code is compliable on 328 selected board.

**Comment by RadekVoltr** [comment]
Next set of changes : 
ARM - changed to Mode as Mode is more usefull and Arm is beeped by copter itself
Data savings for faster updates :
ALT - still 8km maximum but with precision on meters
Batt V,A - over 800V/A maximum
Home Distance - 8km maximum, for distant flights is there commented line for 30b resolution

And I will play with dynamic range detection so (for example) all data in range of -31/31 will be send as one byte regardless of maximum possible number. Can improve update speed with lower than 8100mAh batteries, less than 30 degree angles, lower than 31 V/A voltage and/or current and during close flights.

**Comment by 0n3-70uch** [comment]
Wow nice. Thanks a lot I will test it in a few minutes :-).

**Comment by helmarw** [comment]
update speed is great !!! all sensors getting recorded, great Job !!!
one issue:
there are glitches on the Jetibox. i tried to record a movie (slowmo) so you can see it
https://www.dropbox.com/s/mun4fqy5dmavy8w/20170726_183104000_iOS.MOV?dl=0

again great Job!

PS: I edited the comment, for disarm i was on the wrong screen :) ist was Dis(tance) that should not change of course :)

**Comment by 0n3-70uch** [comment]
Ok, the update speed is very fast now ... BUT it's not working with the REX receiver anymore. I think its a timing problem.

Update: I read that the REX receiver working better with 9800 baudrate. But this is no standard...

**Comment by RadekVoltr** [comment]
Hi guys,

Can you try :
- Change jetispeed to slightly slower
- There is new .h for softserial configuration, you can change interrupt behavior. It could fix Rex issue but slow down updates
- There is FrameISR property on jeti client, can you try to comment assignment to this property? Again, can slow down updates

Thanks
Radek 

Sent from my Windows 10 phone

**Comment by helmarw** [comment]
just a few question because im not sure im getting the right files:
-can you elaborate "how to change the jetispeed" ? 
-the new .h you mentioned is ssconfig.h, or ? there  only one entry (#define ALLOW_ISR)
i quess this is what you mean by commenting FrameISR ?!
-how to change the interrupt behavior in ssconfig.h im also not sure but i dont have a Rex receiver
so i cant test this
thnx
Helmar

**Comment by helmarw** [comment]
i commented "#define ALLOW_ISR" in ssconfig.h
the glitches are gone, but unusable slow update rate
EX-sensors though seem still to work fine

**Comment by RadekVoltr** [comment]
so after few hours of testing different approaches I have another way to test. Clear is that with allowed isr's is timing so bad that softserial simply doesn't work. I tested few ways how to let them work (for example disabling them only during delay and other stuff) and nothing helped. So disabling ISR is needed.

But then I got idea , let's look on old code first :-D and I found there some bugs which had one result - longer wait time between frames. So I got easy idea and here is code :-) Solution work for me , hope that it will work for you too (if not, please try to increase SR params to 4 from 2).

I increased time between jeti frames to 100ms. This 80ms more let FastSerial read enough data to hit mavlink packets at least one per second. With my settings of SR (mentioned up) is possitive hit on one selected packet (roll) in average about 2-3x per second (from 4 sent). It is logical as data send take 100ms and wait is also 100ms so we have 50/50 dutty cycle when we are listening to Mavlink and when not. Number of hits could be increased with UART speed between Arduino and Pixhawk allong with increased frequency of packets. But 57k and 4hz produce acceptable speed for me.

Can you tell me how it works on your setup ? Unfortunatelly all my REXs are in places and on copter I have only RSat2 EX with Mega32U

**Comment by helmarw** [comment]
not bad, update speed is reasonable, but ive to test it outside with GPS.
HDOP shows 0.99, i think it should be 99 but, since i got 0 Satellites and no fix inside
otherwise it looks fine, no glitches on Jetibox ;)
but as i said i need to field test it. sadly im off for a few days so ive to postpone in-depth
test to next week or weekend. Not worth taking the copter up north, bft 5-6 and very high
chance of rain :/ not ideal for any flight ...

Great job !!!

**Comment by 0n3-70uch** [comment]
Hey, very good job. Thanks.

Working nice with the REX receiver and it's very stable and fast :-). The GPS data screen is also working in my case. So cool... I only notice one little thing: the jetibox is a bit slower (changing pages). But this is not so important.

Greetings Fabi

**Comment by RadekVoltr** [comment]
guys, I have little surprise for you. There is new version on git. I fixed HDOP/speed of jetibox screens (see standalone issues) but as my thank you for your cooperation I added something more. Last selected jetibox screen is now stored into eeprom and loaded after each start. So you can add more screens and have selected different screens on different machines. Don't need to have desired easy access stuff on screen number 1 :-)

**Comment by 0n3-70uch** [comment]
I will test it later... thanks for your programming skills and time.

**Comment by helmarw** [comment]
looks smooth and seems to work very well.
what would you suggest to use atmega328 or the Mega32U ?!
i ordered a Pro Micro Modul with ATmega32U4 (basically a Leonardo but as small as the Nano im using right now)
wondering if this make a difference ...
thnx
Helmar

**Comment by 0n3-70uch** [comment]
Hi, same from me: everyhing working fine :-).
